design_documents/accessing_emitted_events.md

```diff
@@ -0,0 +1,218 @@
+    
+    expect_exact_events(array![Event { name: 'StoredName', keys: array![123], data: array![456] }]);
+    let res = contract.emit_store_name(...);  // function has to emit exactly those events defined in the array otherwise it will panic
+    // ...
+}
+```
+
+## `spy_events` cheatcode
+
+Another idea is to give users the highest possible customization. There would be a handler for extracting events emitted
+after the line it was created. Users would have to assert events manually (which could be more complex than the previous
+solution but would add more flexibility).
+
+Introduce `spy_events` cheatcode with the signature:
+
+```cario
+fn spy_events() -> snforge_std::EventSpy
+```
+
+where `snforge_std::EventSpy` would allow for accessing events emitted after its creation.
+`EventSpy` would be defined as follows.
+
+```cario
+struct EventSpy {
+    events: Array<snforge_std::Event>,
+}
+
+struct Event {
+    from: ContractAddress,
+    name: felt252,
+    keys: Array<felt252>,
+    data: Array<felt252>
+}
+
+trait EventFetcher {
+    fn fetch_events(ref self: EventSpy);
+}
+```
+
+Users will be responsible for calling `fetch_events` to load emitted events to the `events` property.
+
+There will also be `assert_emitted` method available on the `EventSpy` struct.
+
+```cairo
+trait EventAssertions {
+    fn assert_emitted(ref self: EventSpy, events: Array<snforge_std::Event>);
+}
+```
+
+It is designed to enable more simplified flow:
+- `fetch_events` will be called internally, so there will always be the newest events,
+- checked events will be removed from the `events` property.
+
+### Usage example
+
+```cario
+// Contract is the same as in the previous example
+
+use snforge_std::spy_events;
+use snforge_std::EventSpy;
+use snforge_std::EventFetcher;
+use snforge_std::EventAsserions;
+
+#[test]
+fn check_emitted_event_simple() {
+    // ...
+	let mut spy = spy_events();  // all events emitted after this line will be saved under the `spy` variable
+    let res = contract.emit_store_name(...);
+    
+    // after this line there will be no events under the `spy.events`
+    spy.assert_emitted(array![Event {from: ..., name: 'StoredName', ...}]);
+
+    let res = contract.emit_store_name(...);
+    let res = contract.emit_store_name(...);
+    
+    spy.assert_emitted(
+        array![
+            Event {from: ..., name: 'StoredName', ...},
+            Event {from: ..., name: 'StoredName', ...}
+        ]
+    );
+    
+    assert(spy.events.len() == 0, 'All events should be consumed');
+    // ...
+}
+
+#[test]
+fn check_emitted_event_complex() {
+    // ...
+	let mut spy = spy_events();  // all events emitted after this line will be saved under the `spy` variable
+    let res = contract.emit_store_name(...);
+    
+    spy.fetch_events();
+    
+    // users can assert events on their own
+    assert(spy.events.len() == 1, 'There should be one event');
+    assert(spy.events.at(0).name == sn_keccak('StoredName'), 'Wrong event name');
```

Comment:
I understand that the names in `events` field will be hashed names, right? Shouldn't we make a separate struct for that then? I know it will be basically the same but with different field name (`hashed_name` prob) but I think it will be less confusing

Reply:
We could. But it is the `felt` in both cases so it is not necessary

Reply:
Of course, these 2 structs would be isomorphic, but I think this way the code describes itself and there would be less confusion

Reply:
The only moment where users have to worry about hashed names is when they are comparing them on their own. In all other cases, it is just a shortstring. I think this is quite clear, but can change it if you think it would be much better. 
I would also add it in the implementation, not in this design doc because it is rather an overview of the planned changes.